test/cri-containerd/clone_test.go

```diff
@@ -0,0 +1,404 @@
+	templatePodID, templateContainerID = createPodAndContainer(ctx, t, client, templateSandboxRequest, templateContainerRequest)
+
+	// Send a context with deadline for waitForTemplateSave function
+	d := time.Now().Add(10 * time.Second)
+	ctx, cancel := context.WithDeadline(ctx, d)
+	defer cancel()
+	waitForTemplateSave(ctx, t, templatePodID)
+	return
+}
+
+// Creates a clone from the given template pod and container.
+// It is the callers responsibility to clean the stop and remove the cloned
+// containers and pods.
+func createClonedContainer(ctx context.Context, t *testing.T, client runtime.RuntimeServiceClient, templatePodID, templateContainerID string, cloneNumber int) (clonedPodID, clonedContainerID string) {
+	cloneSandboxRequest := getClonedPodConfig(cloneNumber, templatePodID)
+	cloneContainerRequest := getClonedContainerConfig(cloneNumber, templateContainerID)
+	clonedPodID, clonedContainerID = createPodAndContainer(ctx, t, client, cloneSandboxRequest, cloneContainerRequest)
+	return
+}
+
+// Runs a command inside given container and verifies if the command executes successfully.
+func verifyContainerExec(ctx context.Context, t *testing.T, client runtime.RuntimeServiceClient, containerID string) {
+	execCommand := []string{
+		"ping",
+		"www.microsoft.com",
+	}
+
+	execRequest := &runtime.ExecSyncRequest{
+		ContainerId: containerID,
+		Cmd:         execCommand,
+		Timeout:     20,
+	}
+
+	r := execSync(t, client, ctx, execRequest)
+	output := strings.TrimSpace(string(r.Stdout))
+	errorMsg := string(r.Stderr)
+	exitCode := int(r.ExitCode)
+
+	if exitCode != 0 || len(errorMsg) != 0 {
+		t.Fatalf("Failed execution inside container %s with error: %s, exitCode: %d", containerID, errorMsg, exitCode)
+	} else {
+		t.Logf("Exec(container: %s) stdout: %s, stderr: %s, exitCode: %d\n", containerID, output, errorMsg, exitCode)
+	}
+}
+
+// A simple test to just create a template container and then create one
+// cloned container from that template.
+func Test_CloneContainer_WCOW(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	client := newTestRuntimeClient(t)
+
+	pullRequiredImages(t, []string{imageWindowsNanoserver})
+
+	templatePodID, templateContainerID := createTemplateContainer(ctx, t, client, getTemplatePodConfig("templatepod"), getTemplateContainerConfig("templatecontainer"))
+	defer removePodSandbox(t, client, ctx, templatePodID)
+	defer stopPodSandbox(t, client, ctx, templatePodID)
+	defer removeContainer(t, client, ctx, templateContainerID)
+	defer stopContainer(t, client, ctx, templateContainerID)
+
+	clonedPodID, clonedContainerID := createClonedContainer(ctx, t, client, templatePodID, templateContainerID, 1)
+	defer removePodSandbox(t, client, ctx, clonedPodID)
+	defer stopPodSandbox(t, client, ctx, clonedPodID)
+	defer removeContainer(t, client, ctx, clonedContainerID)
+	defer stopContainer(t, client, ctx, clonedContainerID)
+
+	verifyContainerExec(ctx, t, client, clonedContainerID)
```

Comment:
Do we expect that we can exec commands in the template container? I assume not, in which case would it be useful to additionally verify that we can't exec in the template container here?

Reply:
We can't exec commands in the template container since it goes into the exited state. I have added checks to verify this state of the template container.